### PR TITLE
fix replay msg ids larger than 128

### PR DESCRIPTION
**Pull request by dayjaby**
### Solved Problem
@dagar When using replays, I found that several topics were not published at all. Digging deep I found that this happened for all uORB topics with a msg_id larger than 128. The old formula behaved weird when casting (signed) char to unsigned int16, so I made the calculation a bit more straightforward.

Before:

```
INFO  [replay] Topic estimator_optical_flow_vel got msg id 125
ERROR [replay] estimator_optical_flow_vel nextDataMessage 96242
INFO  [replay] Topic estimator_optical_flow_vel got msg id 126
ERROR [replay] estimator_optical_flow_vel nextDataMessage 96274
INFO  [replay] Topic estimator_optical_flow_vel got msg id 127
ERROR [replay] estimator_optical_flow_vel nextDataMessage 96306
INFO  [replay] Topic estimator_optical_flow_vel got msg id 65408
ERROR [replay] estimator_optical_flow_vel nextDataMessage 96338
INFO  [replay] Topic estimator_optical_flow_vel got msg id 65409
```
After:

```
INFO  [replay] Topic estimator_optical_flow_vel got msg id 125
ERROR [replay] estimator_optical_flow_vel nextDataMessage 96242
INFO  [replay] Topic estimator_optical_flow_vel got msg id 126
ERROR [replay] estimator_optical_flow_vel nextDataMessage 96274
INFO  [replay] Topic estimator_optical_flow_vel got msg id 127
ERROR [replay] estimator_optical_flow_vel nextDataMessage 96306
INFO  [replay] Topic estimator_optical_flow_vel got msg id 128
ERROR [replay] estimator_optical_flow_vel nextDataMessage 96338
INFO  [replay] Topic estimator_optical_flow_vel got msg id 129
```

### CODE REVIEW

**Comment by dayjaby** [comment]
Tested on a log file to analyze something with battery_status, which did not get published before, but gets correctly published now:

```
$ listener battery_status

TOPIC: battery_status 2 instances

Instance 0:
 battery_status_s
	timestamp: 6855594  (0.194502 seconds ago)
	voltage_v: 46.5778
	voltage_filtered_v: 46.4562
	current_a: 0.0000
	current_filtered_a: 0.0000
	current_average_a: -1.0000
	discharged_mah: 0.0000
	remaining: 1.0000
	scale: 1.0000
	temperature: nan
	cell_count: 12
	low_thr: 0.4000
	crit_thr: 0.3000
	emergen_thr: 0.2000
	voltage_cell_v: [3.8713, 3.8713, 3.8713, 3.8713, 3.8713, 3.8713, 3.8713, 3.8713, 3.8713, 3.8713, 3.8713, 3.8713, 0.0000, 0.0000]
	...
```

File: src/modules/replay/Replay.cpp [hunk 1]
```diff
@@ -387,7 +387,7 @@ Replay::readAndAddSubscription(std::ifstream &file, uint16_t msg_size)
 	_subscription_file_pos = file.tellg();
 
 	uint8_t multi_id = *(uint8_t *)message;
-	uint16_t msg_id = ((uint16_t) message[1]) | (((uint16_t) message[2]) << 8);
+	uint16_t msg_id = static_cast<unsigned char>(message[1]) + (static_cast<unsigned char>(message[2]) << 8);
```
Comment (dagar) [inline src/modules/replay/Replay.cpp:390]: What about not using a `char*` in the first place (`char *message` on line 374)? The underlying data is uint8_t.

``` C++
uint8_t *message = _read_buffer.data();
```

Comment (dayjaby) [reply src/modules/replay/Replay.cpp:390]: Sounds good. I will test that.

**Comment by dayjaby** [comment]
@dagar Found another type issue by using replay:

```
(hrt_elapsed_time(&_manual_control_setpoint.timestamp) < (_param_com_rc_loss_t.get() * 1_s))
```

becomes true IF the timestamp is (slightly) larger than hrt_absolute_time. uint64_t minus uint64_t makes only sense if the first one is always larger than the second one.

Correct formula:
```
        int64_t elapsed = static_cast<int64_t>(hrt_absolute_time()) - static_cast<int64_t>(_manual_control_setpoint.timestamp);   
        PX4_INFO("manual control lost for %" PRId64, elapsed);                                                   
        PX4_INFO("manual control lost for %.2fs", elapsed * 1e-6);
```

resulted in

```
INFO  [commander] manual control lost for -3744
INFO  [commander] manual control lost for -0.00s
```

while the previous formula:
```
        uint64_t elapsed = static_cast<uint64_t>(hrt_absolute_time()) - static_cast<uint64_t>(_manual_control_setpoint.timestamp);
        PX4_INFO("manual control lost for %" PRIu64, elapsed);                                                   
        PX4_INFO("manual control lost for %.2fs", elapsed * 1e-6);                                                    
        PX4_INFO("abs time: %" PRIu64 " man time: %" PRIu64 ", diff: %.6f", hrt_absolute_time(), _manual_control_setpoint.timestamp, (hrt_absolute_time() - _manual_control_setpoint.timestamp) / 1e6);
```

resulted in
```
INFO  [commander] manual control lost for 18446744073709550243
INFO  [commander] manual control lost for 18446744073709.55s
INFO  [commander] abs time: 245735 man time: 247108, diff: 18446744073709.546875
```

Imho a time diff shouldn't be unsigned, but signed.

**Comment by dagar** [comment]
> @dagar Found another type issue by using replay:

I'm not sure, in general I think it's reasonable for a system to expect its monotonic time source is a monotonic time source and as long as we're populating the msg timestamp directly with a fresh hrt_absolute_time() it should be a safe comparison for any downstream consumer.

All that is to say, I propose we address it as `replay` issue rather than something more fundamental.

This is what I had in mind for a replay refactor.
 - the first field in every msg is a timestamp that should be populated as publication metadata (I was considering pushing this into uORB itself to prevent any potential misuse)
 - in the `replay` module we build an a timestamp index for all logged msgs
 - when actually replaying we step through logged msgs one at a time with lockstep enabled setting the time for each step from our current msg timestamp

Thoughts? If you're interested I suspect we could hack this together pretty quickly. I'm mainly interested in the version of replay where we push through all relevant data as fast as possible, but if desired I think we could also add playback speed.

**Comment by dayjaby** [comment]
> as long as we're populating the msg timestamp directly with a fresh hrt_absolute_time()

In the current state, this is not being done. There is some sleep-mechanic, which sometimes causes a uORB message to be published without the hrt to be reset. I guess if we completely remove the sleep, the problem should be already solved.

> the first field in every msg is a timestamp that should be populated as publication metadata (I was considering pushing this into uORB itself to prevent any potential misuse)

You mean remove timestamp from the msg/*.msg files and instead enforce it? Sounds good.

> I'm mainly interested in the version of replay where we push through all relevant data as fast as possible

That also aligns with my use case: Basically I want to use replay for some unit tests/test-driven design: 
1. Flight abc.ulg had some issue with the main navigation state in certain edge cases (thus my focus on replaying `commander` right now)
2. Write a unit test that tests against the expected behaviour
3. The fix for the issue in 1. (and in the future, any other modifications to the module under test) are supposed to pass the unit test

**Comment by dagar** [comment]
@dayjaby I'm away for a few days, but when I'm back do you want to schedule a call to discuss and actually get this done? 

In the meantime should we get the msg id fix in?

**Comment by dayjaby** [comment]
> In the meantime should we get the msg id fix in?

I'd appreciate that!

File: src/modules/replay/Replay.cpp (hunk 1 above)
Comment (dayjaby) [reply src/modules/replay/Replay.cpp]: Tested and still works with IDs >= 128. Force-pushed this.

**Review by dagar — APPROVED** [review]
